src/main/java/com/fulcrumgenomics/util/PickLongIndices.scala

```diff
@@ -0,0 +1,291 @@
+      val string = new String(index)
+
+      // structure checking (worstDeltaG) is done here instead of in randomAcceptableIndex because,
+      // even though it's a per-index property, it's much slower than all the other checks, and
+      // when making many attempts to find a suitable next index, the time spent structure checking
+      // all the indices we're going to discard for too-few-edits anyway explodes.
+      if ( !picks.contains(string) &&
+           (allowReverses || !picks.contains(string.reverse)) &&
+           (allowReverseComplements || !picks.contains(SequenceUtil.reverseComplement(string))) &&
+           picks.forall(p => mismatches(index, p) >= minEdits) &&
+           worstDeltaG(index).forall(_ >= minDeltaG)) {
+        pick = Some(index)
+      }
+      attempt += 1
+    }
+
+    pick
+  }
+
+  /** Generates the next random index which is acceptable as a standalone index, prior to any secondary structure checking. */
+  @tailrec private def randomAcceptableIndex(length: Int): Index = {
+    val index  = randomIndex(length)
+    val string = new String(index)
+    val gc = SequenceUtil.calculateGc(index)
+
+    if (gc >= minGc && gc <= maxGc &&
+        homopolymerLengthOk(index, maxHomopolymer) &&
+        !AvoidKmers.contains(string) &&
+        (allowPalindromes || !isPalindrome(index))
+    ) {
+      index
+    }
+    else {
+      randomAcceptableIndex(length)
+    }
+  }
+
+  /** Returns true if a sequence is a palindrome, otherwise false. */
+  private[util] def isPalindrome(seq: Index): Boolean = {
+    var i = 0
+    var j = seq.length-1
+    var palindrome = true
+    while (palindrome && i < j) {
+      palindrome = SequenceUtil.basesEqual(seq(i), SequenceUtil.complement(seq(j)))
+      i += 1
+      j -= 1
+    }
+    palindrome
+  }
+
+  /** If we have structure prediction parameters, returns Some(min deltaG) else None. */
+  private[util] def worstDeltaG(index: Index): Option[Double] = {
+    val sIndex = new String(index)
+    dnaFoldPredictor.map(folder => adapters.map(a => folder.predict(a.replaceAll("N+", sIndex)).deltaG()).min)
+  }
+
+  /** Generates a random index sequence of the desired length. */
+  private def randomIndex(length: Int): Index = {
+    val index = new Array[Byte](length)
+    forloop (0, length) { i => index(i) = Bases(random.nextInt(4)) }
+    index
+  }
+
+  /** Checks to see if the index sequence contains any homopolymers longer than longest. */
+  private[util] def homopolymerLengthOk(index: Index, longest: Int): Boolean = {
+    var end = index.length - longest
+    var ok = true
+
+    // Check for a homopolymer starting at each base up to longest+1 from the end of the index
+    forloop (0)(ok && _ < end)(_ + 1) { i =>
+      val base = index(i)
+      var j = i+1
+      var inHomoPolymer = true
+      forloop (i+1)(inHomoPolymer && _ < i+longest+1)(_ + 1) { j =>
+        if (base != index(j)) inHomoPolymer = false
+      }
+
+      ok = !inHomoPolymer
```

Comment:
It's true, but then I think I'd have to track homopolyerLength through the second loop.  I think I prefer the simpler method and marginally slower runtime.